### ISSUE TITLE
Bug: [unified-signatures] Does not differentiate truly private methods.

### ISSUE THREAD

**Issue by LukeAbby** [typescript-eslint/typescript-eslint#10800]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=CYUwxgNghgTiAEkoGdnwKIA8oFsAOECA3gLABQ88yALlNQJZjwDEeM9AbnSALIjUALAPbAAjAAooALngA7AK44ARiBgBKGRyH1gAbnKUadRizadufQSIBMkmQuWqANPCUB%2BGTXayA5hvhaOvpkAL5AA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFETRoAe2iRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class Example {
  static #privateMethod1(a: number): void;
  static #privateMethod2(a: number, b?: string): void;
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/unified-signatures": "error"
  }
}
```

### tsconfig

```jsonc

```

### Expected Result

No error.

### Actual Result

> These overloads can be combined into one signature with an optional parameter.
